### ISSUE TITLE
OutputTableRetentionDays Not Funtioning Properly 

### ISSUE THREAD

**Issue by tonydba** [BrentOzarULTD/SQL-Server-First-Responder-Kit#3405]
Version of the script
SELECT @Version = '8.17', @VersionDate = '20231010';.

What is the current behaviour?
I scheduled sp_BlitzFirst to run hourly and log data in the backend tables. I have set retention as 2 days as shown using OutputTableRetentionDays. But it is keeping 4 days of data on backend tables.

If the current behaviour is a bug, please provide the steps to reproduce.
Run the job with OutputTableRetentionDays=2

What is the expected behavior?
Retain data for only 2 days and not 4 days .

Which versions of SQL Server and which OS are affected by this issue? Did this work in previous versions of our procedures?
Microsoft SQL Server 2019 (RTM-CU22-GDR) (KB5029378) - 15.0.4326.1 (X64) Aug 18 2023 14:05:15 Copyright (C) 2019 Microsoft Corporation Developer Edition (64-bit) on Windows Server 2019 Standard 10.0 (Build 17763: ) (Hypervisor)
Table_Retention_Checks.docx

**Comment by tonydba** [comment]
[Table_Retention_Checks.docx](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/files/13601311/Table_Retention_Checks.docx)

**Comment by tonydba** [comment]
Below are the parameters I am using for the  hourly job. I haven't made any changes to underlying sp_Blitz* procedures. 

exec sp_BlitzFirst
@OutputDatabaseName = 'DBA',
@OutputSchemaName = 'dbo',
@OutputTableName = 'BlitzFirst',
@OutputTableNameFileStats= 'BlitzFirst_FileStats',
@OutputTableNamePerfmonStats = 'BlitzFirst_PerfmonStats',
@OutputTableNameWaitStats = 'BlitzFirst_WaitStats',
@OutputTableNameBlitzCache = 'BlitzCache',
@OutputTableNameBlitzWho = 'BlitzWho', 
@OutputTableRetentionDays = 2 -- Added for housekeeping 

Regards,
Tony

**Comment by erikdarlingdata** [comment]
@tonydba Is this happening across all tables, or just the two in your screenshot?

**Comment by tonydba** [comment]
[Table Retention Checks--ALL--12082023.docx](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/files/13619023/Table.Retention.Checks--ALL--12082023.docx)

**Comment by tonydba** [comment]
Th output for all tables mentioned in previous output arebelow ? so issue is across all tables sp_BlitzFirst populates . 

SELECT DISTINCT(CONVERT(date, checkDate)) from  [dbo].[Blitz***]
2023-12-05
2023-12-06
2023-12-07
2023-12-08


Regards,
Tony

**Comment by BrentOzar** [comment]
@tonydba here's a repro script you can run in your own environment to see whether the parameter is working correctly:

```
/* Work in a separate database to avoid breaking your production monitoring: */
CREATE DATABASE RetentionTesting;
GO
USE RetentionTesting;
GO

DROP TABLE IF EXISTS dbo.BlitzFirst
DROP TABLE IF EXISTS dbo.BlitzFirst_FileStats
DROP TABLE IF EXISTS dbo.BlitzFirst_PerfmonStats
DROP TABLE IF EXISTS dbo.BlitzFirst_WaitStats
DROP TABLE IF EXISTS dbo.BlitzCache
DROP TABLE IF EXISTS dbo.BlitzWho
GO

EXEC sp_BlitzFirst @OutputDatabaseName = 'RetentionTesting',
@OutputSchemaName = 'dbo',
@OutputTableName = 'BlitzFirst',
@OutputTableNameFileStats = 'BlitzFirst_FileStats',
@OutputTableNamePerfmonStats = 'BlitzFirst_PerfmonStats',
@OutputTableNameWaitStats = 'BlitzFirst_WaitStats',
@OutputTableNameBlitzCache = 'BlitzCache',
@OutputTableNameBlitzWho = 'BlitzWho'
GO

/* Back-date all of the data 5 days: */
UPDATE dbo.BlitzFirst SET CheckDate = DATEADD(dd, -5, CheckDate);
UPDATE dbo.BlitzFirst_FileStats SET CheckDate = DATEADD(dd, -5, CheckDate);
UPDATE dbo.BlitzFirst_PerfmonStats SET CheckDate = DATEADD(dd, -5, CheckDate);
UPDATE dbo.BlitzFirst_WaitStats SET CheckDate = DATEADD(dd, -5, CheckDate);
UPDATE dbo.BlitzCache SET CheckDate = DATEADD(dd, -5, CheckDate);
UPDATE dbo.BlitzWho SET CheckDate = DATEADD(dd, -5, CheckDate);
GO

/* Run it with a 1-day retention policy: */
EXEC sp_BlitzFirst @OutputDatabaseName = 'RetentionTesting',
@OutputSchemaName = 'dbo',
@OutputTableName = 'BlitzFirst',
@OutputTableNameFileStats = 'BlitzFirst_FileStats',
@OutputTableNamePerfmonStats = 'BlitzFirst_PerfmonStats',
@OutputTableNameWaitStats = 'BlitzFirst_WaitStats',
@OutputTableNameBlitzCache = 'BlitzCache',
@OutputTableNameBlitzWho = 'BlitzWho',
@OutputTableRetentionDays = 1;
GO

/* Look for data older than 1 day: */
SELECT * FROM dbo.BlitzFirst WHERE CheckDate < DATEADD(dd, -1, GETDATE())
SELECT * FROM dbo.BlitzFirst_FileStats WHERE CheckDate < DATEADD(dd, -1, GETDATE())
SELECT * FROM dbo.BlitzFirst_PerfmonStats WHERE CheckDate < DATEADD(dd, -1, GETDATE())
SELECT * FROM dbo.BlitzFirst_WaitStats WHERE CheckDate < DATEADD(dd, -1, GETDATE())
SELECT * FROM dbo.BlitzCache WHERE CheckDate < DATEADD(dd, -1, GETDATE())
SELECT * FROM dbo.BlitzWho WHERE CheckDate < DATEADD(dd, -1, GETDATE())
GO
```

Here are the results from my system showing the data was deleted correctly:
![Screenshot 2023-12-08 at 9 52 37 AM](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/assets/245462/7499f9e2-4fd3-4aa2-beaa-2c6ca99fb9aa)

Can you run this test on your system, and report back if it's working successfully or not? If it's not giving you the results you'd expect, feel free to modify the above repro script to demonstrate the issue that you're having. Thanks!

**Comment by tonydba** [comment]
Here is the output from the new DB created for testing retention. I can see it's working as expected here. 

![image](https://github.com/BrentOzarULTD/SQL-Server-First-Responder-Kit/assets/125384726/5c548964-5313-4829-9655-f5ccbe074dd7)

**Comment by tonydba** [comment]
We still have an issue with our original database. Btw.. would you please answer below query: 

Which is a good practice which you recommend? 

1.  Leave the underlying procedures untouched and then control the retention explicitly by passing parameter in job (@OutputTableRetentionDays = ) .  See below for an example.

exec sp_BlitzFirst
@OutputDatabaseName = 'DBA',
@OutputSchemaName = 'dbo',
@OutputTableName = 'BlitzFirst',
@OutputTableNameFileStats= 'BlitzFirst_FileStats',
@OutputTableNamePerfmonStats = 'BlitzFirst_PerfmonStats',
@OutputTableNameWaitStats = 'BlitzFirst_WaitStats',
@OutputTableNameBlitzCache = 'BlitzCache',
@OutputTableNameBlitzWho = 'BlitzWho', 
@OutputTableRetentionDays = 3 -- Added for housekeeping 

2.  Hardcode the retention in the underlying codes for procedures by altering the default procedures.  
@OutputTableRetentionDays TINYINT = 3 ,-- changed from 7

Regards,
Tony

**Comment by BrentOzar** [comment]
@tonydba I can't troubleshoot your particular database here on Github, but if you'd like to hire me for personalized consulting advice, feel free to go to BrentOzar.com and click Consulting at the top of the screen. Hope that's fair.

**Comment by tonydba** [comment]
I agree. It would be great if you can answer my last query on the usage of OutputTableRetentionDays

**Comment by BrentOzar** [comment]
Tony, I feel like I’ve been pretty fair in providing the scripts, helping you troubleshoot after you reported a bug that doesn’t seem to exist, and pointing you to resources for personalized help. I’m going to draw the line here. Thanks for understanding.---Tiny glass keyboardTypos flowing like riversAs winter snow thawsOn Dec 8, 2023, at 10:35 AM, tonydba ***@***.***> wrote:﻿
I agree. It would be great if you can answer my last query on the usage of OutputTableRetentionDays
